x/cdp/keeper/cdp.go

```diff
@@ -0,0 +1,407 @@
+	store := prefix.NewStore(ctx.KVStore(k.key), types.CdpIDKeyPrefix)
+	cdpIDs, found := k.GetCdpIdsByOwner(ctx, cdp.Owner)
+	if !found {
+		return
+	}
+	updatedCdpIds := []uint64{}
+	for _, id := range cdpIDs {
+		if id != cdp.ID {
+			updatedCdpIds = append(updatedCdpIds, id)
+		}
+	}
+	if len(updatedCdpIds) == 0 {
+		store.Delete(cdp.Owner)
+	}
+	store.Set(cdp.Owner, k.cdc.MustMarshalBinaryLengthPrefixed(updatedCdpIds))
+
+}
+
+// IndexCdpByCollateralRatio sets the cdp id in the store, indexed by the collateral type and collateral to debt ratio
+func (k Keeper) IndexCdpByCollateralRatio(ctx sdk.Context, denom string, id uint64, collateralRatio sdk.Dec) {
+	store := prefix.NewStore(ctx.KVStore(k.key), types.CollateralRatioIndexPrefix)
+	db, _ := k.GetDenomPrefix(ctx, denom)
+	store.Set(types.CollateralRatioKey(db, id, collateralRatio), types.GetCdpIDBytes(id))
+}
+
+// RemoveCdpCollateralRatioIndex deletes the cdp id from the store's index of cdps by collateral type and collateral to debt ratio
+func (k Keeper) RemoveCdpCollateralRatioIndex(ctx sdk.Context, denom string, id uint64, collateralRatio sdk.Dec) {
+	store := prefix.NewStore(ctx.KVStore(k.key), types.CollateralRatioIndexPrefix)
+	db, _ := k.GetDenomPrefix(ctx, denom)
+	store.Delete(types.CollateralRatioKey(db, id, collateralRatio))
+}
+
+// GetDebtDenom returns the denom of debt in the system
+func (k Keeper) GetDebtDenom(ctx sdk.Context) (denom string) {
+	store := prefix.NewStore(ctx.KVStore(k.key), types.DebtDenomKey)
+	bz := store.Get([]byte{})
+	k.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &denom)
+	return
+}
+
+// SetDebtDenom set the denom of debt in the system
+func (k Keeper) SetDebtDenom(ctx sdk.Context, denom string) {
+	if denom == "" {
+		panic("debt denom not set in genesis")
+	}
+	store := prefix.NewStore(ctx.KVStore(k.key), types.DebtDenomKey)
+	store.Set([]byte{}, k.cdc.MustMarshalBinaryLengthPrefixed(denom))
+	return
+}
+
+// ValidateCollateral validates that a collateral is valid for use in cdps
+func (k Keeper) ValidateCollateral(ctx sdk.Context, collateral sdk.Coins) sdk.Error {
+	if len(collateral) != 1 {
+		return types.ErrInvalidCollateralLength(k.codespace, len(collateral))
+	}
+	_, found := k.GetCollateral(ctx, collateral[0].Denom)
+	if !found {
+		return types.ErrCollateralNotSupported(k.codespace, collateral[0].Denom)
+	}
+	return nil
+}
+
+// ValidatePrincipalAdd validates that an asset is valid for use as debt when creating a new cdp
+func (k Keeper) ValidatePrincipalAdd(ctx sdk.Context, principal sdk.Coins) sdk.Error {
+	for _, dc := range principal {
+		dp, found := k.GetDebt(ctx, dc.Denom)
+		if !found {
+			return types.ErrDebtNotSupported(k.codespace, dc.Denom)
+		}
+		if sdk.NewCoins(dc).IsAnyGT(dp.DebtLimit) {
+			return types.ErrExceedsDebtLimit(k.codespace, sdk.NewCoins(dc), dp.DebtLimit)
+		}
+		if dc.Amount.LT(dp.DebtFloor) {
+			return types.ErrBelowDebtFloor(k.codespace, sdk.NewCoins(dc), dp.DebtFloor)
+		}
```

Comment:
Should a DebtFloor check also be added somewhere in `DepositCollateral()`/`ValidatePrincipalDraw()` to prevent dust with additional principal coin types? May allow us to combine this function with `ValidatePrincipalDraw()`, but I'm not sure I fully understand all the potential implications of such an addition.